samples/swift/uidemo/ChatViewController.swift

```diff
@@ -0,0 +1,251 @@
+          let colors = ChatCollectionViewCell.othersColors
+          cell.containerView.backgroundColor = colors.background
+          cell.textLabel.textColor = colors.text
+          cell.leadingConstraint.active = false
+          cell.trailingConstraint.constant = constant
+          cell.trailingConstraint.active = true
+        }
+      }
+      
+      // Unfortunately neither FirebaseArray nor FirebaseCollectionViewDataSource
+      // provide an event listener, so in order to scroll on new insertions we
+      // still need to use the query directly.
+      self.query!.observeEventType(.ChildAdded, withBlock: { [unowned self] _ in
+        self.scrollToBottom(animated: true)
+      })
+    }
+    
+    // Notification boilerplate to handle keyboard appearance/disappearance
+    NSNotificationCenter.defaultCenter().addObserver(self,
+                                                     selector: #selector(keyboardWillShow),
+                                                     name: UIKeyboardWillShowNotification,
+                                                     object: nil)
+    NSNotificationCenter.defaultCenter().addObserver(self,
+                                                     selector: #selector(keyboardWillHide),
+                                                     name: UIKeyboardWillHideNotification,
+                                                     object: nil)
+  }
+  
+  override func viewWillDisappear(animated: Bool) {
+    super.viewWillDisappear(animated)
+    NSNotificationCenter.defaultCenter().removeObserver(self)
+  }
+  
+  deinit {
+    NSNotificationCenter.defaultCenter().removeObserver(self)
+  }
+  
+  @objc private func keyboardWillShow(notification: NSNotification) {
+    let userInfo = notification.userInfo!
+    let endFrameValue = userInfo[UIKeyboardFrameEndUserInfoKey] as! NSValue
+    let endHeight = endFrameValue.CGRectValue().size.height
+    
+    self.bottomConstraint.constant = endHeight
+    
+    let curve = UIViewAnimationCurve(rawValue: userInfo[UIKeyboardAnimationCurveUserInfoKey] as! Int)!
+    let duration = userInfo[UIKeyboardAnimationDurationUserInfoKey] as! Double
+    
+    UIView.setAnimationCurve(curve)
+    UIView.animateWithDuration(duration) { 
+      self.view.layoutIfNeeded()
+    }
+  }
+  
+  @objc private func keyboardWillHide(notification: NSNotification) {
+    self.bottomConstraint.constant = 6
+    
+    let userInfo = notification.userInfo!
+    let curve = UIViewAnimationCurve(rawValue: userInfo[UIKeyboardAnimationCurveUserInfoKey] as! Int)!
+    let duration = userInfo[UIKeyboardAnimationDurationUserInfoKey] as! Double
+    
+    UIView.setAnimationCurve(curve)
+    UIView.animateWithDuration(duration) {
+      self.view.layoutIfNeeded()
+    }
+  }
+  
+  @objc private func didTapSend(sender: AnyObject) {
+    guard let user = self.auth?.currentUser else { return }
+    let uid = user.uid
+    let name = "User " + (uid as NSString).substringToIndex(6)
```

Comment:
This is the lazy (bad) way of getting substrings by disregarding Unicode. Swift's substring thing is much more verbose for Unicode reasons.

```swift
let other = str[str.characters.startIndex..<str.characters.startIndex.advancedBy(6)]
```

This code behaves subtly differently than the code in the sample, and I have no preference as to which one we keep. Should I replace the NSString cast with the Swift string subscript?